File: 413.等差数列划分.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& A) {
        if(A.size() <= 2)
            return 0;
        int n = A.size();

        // int* dp = new int[n]();
        // for(int i = 2; i < n; ++i){
        //     if(A[i] - A[i - 1] == A[i - 1] - A[i - 2])
        //         dp[i] = dp[i - 1] + 1;
        // }
        // int total = 0;
        // for(int i = 0; i < n; ++i)
        //     total += dp[i];

        int add = 0;
        int total = 0;
        for(int i = 2; i < n; ++i){
            if(A[i] - A[i - 1] == A[i - 1] - A[i - 2])
                total += ++add;
            else
                add = 0;
        }

        return total;
    }
};

File: 动态规划dp/413.等差数列划分.js
/**
 * @param {number[]} nums
 * @return {number}
 */
// dp[i] 表示包含 nums[i] 的等差数列个数
var numberOfArithmeticSlices = function (nums) {
    let n = nums.length;
    if (n <= 2)
        return 0;
    let dp = new Array(n).fill(0);
    let total = 0;
    for (let i = 2; i < n; ++i) {
        if (nums[i] - nums[i - 1] === nums[i - 1] - nums[i - 2]) {
            dp[i] = dp[i - 1] + 1;
        }
        total += dp[i];
    }

    return total;
};

var numberOfArithmeticSlices = function (nums) {
    let total = 0,
        add = 0;
    for (let i = 2; i < n; ++i) {
        if (nums[i] - nums[i - 1] === nums[i - 1] - nums[i - 2]) {
            ++add;
            total += add;
        } else
            add = 0;
    }
    return total;
}